### PR TITLE
feat: add shared Zod validators

### LINKED ISSUE #4127: [Streamline - architecture] Standardize Input Validation Patterns
## Problem Statement
Validation is scattered across resolvers/TypeBox; duplicated patterns for IDs, dates, pagination.

## What needs to be done
Create reusable Zod validators; refactor GraphQL inputs and REST schemas to use them (two PRs).

## Approach
- PR1: shared validator utilities + tests.
- PR2: refactor selected inputs/resolvers to use shared validators; update docs.

## PR / Files constraint
- Each PR SHOULD touch 5–20 files.

## Files (scope guidance)
- PR1: `src/graphql/validators/*`, `test/graphql/validators/*`
- PR2: `src/graphql/inputs/*`, selected resolvers/routes schemas, docs

## Links
- Blocked by: #4126
- Blocks: None
- Phase: Phase 3  

## Sample code
```ts
// src/graphql/validators/idValidators.ts
import { z } from "zod";
export const uuidSchema = z.string().uuid("Must be a valid UUID");
export const idSchema = z.string().min(1, "ID is required");

// Example usage in an input
export const createPostInputSchema = z.object({
  organizationId: uuidSchema,
  title: z.string().min(1).max(200),
});
```

---
- **AI Generated:** Provided only as a guide. This is a representation of what could be done
---

Examples of possible ways to introduce reusable Zod validators, thin parse helpers for GraphQL and REST that integrate with our unified error handling, plus examples and minimal tests.

Core validator library (Zod)
```ts
// src/graphql/validators/core.ts
import { z } from "zod";

// Basic scalars
export const trimmedString = z.string().transform((s) => s.trim());
export const nonEmptyString = trimmedString.min(1, "Must not be empty");
export const uuid = z.string().uuid("Must be a valid UUID");
export const ulid = z.string().regex(/^[0-9A-HJKMNP-TV-Z]{26}$/, "Must be a valid ULID");
export const email = trimmedString.email("Must be a valid email");
export const url = trimmedString.url("Must be a valid URL");

// Date/time
export const isoDateString = trimmedString
  .regex(/^\d{4}-\d{2}-\d{2}$/, "Must be YYYY-MM-DD");
export const isoDateTimeString = trimmedString
  .regex(/^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:\.\d+)?Z$/, "Must be ISO-8601 UTC datetime (Z)");
export const coerceDate = z.coerce.date({ required_error: "Date required" });

// IDs (GraphQL ID is a string — keep semantics explicit at resolver-level)
export const id = nonEmptyString;               // GraphQL ID (opaque)
export const orgId = uuid.describe("Organization UUID");
export const userId = uuid.describe("User UUID");

// Enums / common values
export const sortOrder = z.enum(["asc", "desc"]);
export const postSort = z.enum(["NEW", "TOP"]);

// Pagination
export const pagination = z.object({
  limit: z
    .number({ coerce: true })
    .int("limit must be an integer")
    .min(1, "limit must be >= 1")
    .max(100, "limit must be <= 100")
    .default(20),
  cursor: trimmedString.nullish().transform((v) => v ?? null),
});
export type Pagination = z.infer<typeof pagination>;

// Common input shapes
export const organizationIdArg = z.object({ id: orgId });

export const createPostInput = z.object({
  organizationId: orgId,
  title: nonEmptyString.max(200, "Title too long"),
  body: trimmedString.max(10_000, "Body too long").optional().default(""),
  tags: z.array(nonEmptyString.max(32)).max(20).optional().default([]),
  visibility: z.enum(["public", "private"]).default("public"),
});

export const eventRangeInput = z.object({
  organizationId: orgId,
  from: isoDateTimeString.optional(),
  to: isoDateTimeString.optional(),
  onlyPublic: z.boolean().optional().default(false),
});
```

Parse helpers (GraphQL + REST) that map Zod errors to unified errors
```ts
// src/graphql/validators/helpers.ts
import type { z } from "zod";
import { TalawaGraphQLError } from "~/src/utilities/TalawaGraphQLError";
import { ErrorCode } from "~/src/utilities/errors/errorCodes";

export async function zParseOrThrow<TSchema extends z.ZodTypeAny>(
  schema: TSchema,
  data: unknown
): Promise<z.infer<TSchema>> {
  const res = await schema.safeParseAsync(data);
  if (!res.success) {
    throw new TalawaGraphQLError({
      code: ErrorCode.INVALID_ARGUMENTS,
      message: "Invalid arguments",
      details: res.error.flatten(),
    });
  }
  return res.data;
}
```

```ts
// src/routes/validation/zodReply.ts
import type { FastifyReply } from "fastify";
import type { z } from "zod";
import { normalizeError } from "~/src/utilities/errors/errorTransformer";
import { ErrorCode } from "~/src/utilities/errors/errorCodes";

export async function zReplyParsed<TSchema extends z.ZodTypeAny>(
  reply: FastifyReply,
  schema: TSchema,
  data: unknown
): Promise<z.infer<TSchema> | undefined> {
  const parsed = await schema.safeParseAsync(data);
  if (!parsed.success) {
    const norm = normalizeError(parsed.error); // maps to invalid_arguments + 400
    return reply.status(400).send({
      error: {
        code: ErrorCode.INVALID_ARGUMENTS,
        message: "Invalid request body",
        details: parsed.error.flatten(),
        correlationId: reply.request.id,
      },
    }) as unknown as undefined;
  }
  return parsed.data;
}
```

GraphQL usage examples
```ts
// src/graphql/types/Mutation/createPost.ts (example)
import type { GraphQLContext } from "~/src/graphql/context";
import { createPostInput } from "~/src/graphql/validators/core";
import { zParseOrThrow } from "~/src/graphql/validators/helpers";

export const Mutation_createPost = {
  resolve: async (_p: unknown, args: { input: unknown }, ctx: GraphQLContext) => {
    const input = await zParseOrThrow(createPostInput, args.input);

    // ... domain checks (authz, org visibility) ...
    const post = await ctx.drizzleClient.insert(ctx.drizzleClient.tables.postsTable)
      .values({
        organizationId: input.organizationId,
        title: input.title,
        body: input.body,
      })
      .returning()
      .then((r) => r[0]);

    return post;
  },
};
```

```ts
// src/graphql/types/Query/eventFeed.ts (example)
import type { GraphQLContext } from "~/src/graphql/context";
import { eventRangeInput, pagination } from "~/src/graphql/validators/core";
import { zParseOrThrow } from "~/src/graphql/validators/helpers";

export const Query_eventFeed = {
  resolve: async (_p: unknown, args: { input: unknown; page?: unknown }, ctx: GraphQLContext) => {
    const filter = await zParseOrThrow(eventRangeInput, args.input);
    const page = await zParseOrThrow(pagination, args.page ?? {});
    const from = filter.from ? new Date(filter.from) : new Date(Date.now() - 7 * 24 * 3600_000);
    const to   = filter.to   ? new Date(filter.to)   : new Date(Date.now() + 30 * 24 * 3600_000);

    const rows = await ctx.drizzleClient.query.eventsTable.findMany({
      where: (f, op) => op.and(
        op.eq(f.organizationId, filter.organizationId),
        op.gte(f.startAt, from),
        op.lte(f.endAt, to),
        filter.onlyPublic ? op.eq(f.isPublic, true) : op.gte(f.isPublic, false),
      ),
      orderBy: (t, op) => op.asc(t.startAt),
      limit: page.limit,
    });

    const nextCursor = rows.length === page.limit ? rows[rows.length - 1].id : null;
    return { edges: rows, pageInfo: { nextCursor } };
  },
};
```

REST usage example (Fastify + Zod)
```ts
// src/routes/organizations.ts
import type { FastifyInstance } from "fastify";
import { z } from "zod";
import { uuid, pagination } from "~/src/graphql/validators/core";
import { zReplyParsed } from "~/src/routes/validation/zodReply";

const listArgs = z.object({
  q: z.string().optional().default(""),
  page: pagination.optional().default({ limit: 20, cursor: null }),
});

const organizationIdParam = z.object({ id: uuid });

export default async function organizationsRoutes(app: FastifyInstance) {
  app.get("/organizations", async (req, reply) => {
    const parsed = await zReplyParsed(reply, listArgs, { ...req.query, page: (req.query as any)?.page });
    if (!parsed) return;
    const { q, page } = parsed;

    const rows = await app.drizzleClient.query.organizationsTable.findMany({
      where: (f, op) => q ? op.like(f.name, `%${q}%`) : op.gte(f.name, ""),
      limit: page.limit,
    });

    reply.send({ data: rows });
  });

  app.get("/organizations/:id", async (req, reply) => {
    const params = await zReplyParsed(reply, organizationIdParam, req.params);
    if (!params) return;
    const org = await app.drizzleClient.query.organizationsTable.findFirst({
      where: (f, op) => op.eq(f.id, params.id),
    });
    if (!org) return reply.status(404).send({ error: { code: "not_found", message: "Organization not found" } });
    reply.send({ data: org });
  });
}
```

Validator-focused tests (keep schemas directly testable)
```ts
// test/graphql/validators/pagination.test.ts
import { describe, it, expect } from "vitest";
import { pagination } from "~/src/graphql/validators/core";

describe("pagination validator", () => {
  it("defaults limit and cursor", async () => {
    const out = await pagination.parseAsync({});
    expect(out.limit).toBe(20);
    expect(out.cursor).toBeNull();
  });

  it("coerces numeric strings and clamps", async () => {
    const out = await pagination.parseAsync({ limit: "100" });
    expect(out.limit).toBe(100);
    await expect(pagination.parseAsync({ limit: 0 })).rejects.toBeTruthy();
    await expect(pagination.parseAsync({ limit: 101 })).rejects.toBeTruthy();
  });
});
```

```ts
// test/graphql/validators/ids.test.ts
import { describe, it, expect } from "vitest";
import { uuid, id } from "~/src/graphql/validators/core";

describe("id validators", () => {
  it("uuid accepts valid UUID", () => {
    expect(() => uuid.parse("550e8400-e29b-41d4-a716-446655440000")).not.toThrow();
  });
  it("id rejects empty strings", () => {
    expect(() => id.parse("   ")).toThrow();
  });
});
```

Pattern checklist to standardize across the codebase
- Always define Zod schemas in src/graphql/validators/* and export them for direct unit testing.
- In GraphQL resolvers, validate args with zParseOrThrow(schema, args) before business logic; this unifies error shape via TalawaGraphQLError with code invalid_arguments.
- In REST routes, validate with zReplyParsed(reply, schema, data). On failure, return 400 with the unified error format from your error handler.
- Prefer trimmedString/nonEmptyString to avoid hidden whitespace bugs.
- Keep pagination defaults and bounds (limit ∈ [1, 100]) in one place (pagination schema).
- Use uuid/orgId/userId everywhere to avoid ad-hoc regex checks.
- For viewer-dependent queries (e.g., visibility), keep that logic separate from validation; but ensure required flags (onlyPublic) are present and typed.
- When a resolver/route needs a new pattern, add it to core.ts and unit test it under test/graphql/validators/*.

### PR DESCRIPTION
**Pull request by aadithya2112**
**What kind of change does this PR introduce?**
Feature - Shared validation utilities

**Issue Number:**

Fixes #4127  

**Snapshots/Videos:**

**If relevant, did you update the documentation?**

**Summary**
This PR introduces reusable Zod validation utilities to standardize input validation across GraphQL and REST APIs. This is Part 1 of 2 for issue 4127.

Added:
- Core validators for IDs, dates, pagination, and common patterns
- GraphQL helper zParseOrThrow for resolver-level validation
- REST helper zReplyParsed for Fastify route validation
- Comprehensive test suite with 89 tests (100% pass rate)

**Does this PR introduce a breaking change?**
No

### CodeRabbit AI Review
- [ ] I have reviewed and addressed all critical issues flagged by CodeRabbit AI
- [ ] I have implemented or provided justification for each non-critical suggestion
- [ ] I have documented my reasoning in the PR comments where CodeRabbit AI suggestions were not implemented

### Test Coverage
- [x] I have written tests for all new changes/features
- [x] I have verified that test coverage meets or exceeds 95%
- [x] I have run the test suite locally and all tests pass

**Other information**

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-api/blob/master/CONTRIBUTING.md)?**

Yes

## Summary by CodeRabbit

* **New Features**
  * OAuth account linking mutation; venue creation now returns createdAt.
  * Centralized, reusable input validators and unified validation helpers for GraphQL and REST producing consistent, structured validation errors.

* **API Changes**
  * Query.userTags updated: list nullability tightened and descriptive docs improved.

* **Tests**
  * Extensive test suites added covering validators, helpers, REST reply handling, error shapes, defaults, and edge cases.

### CODE REVIEW

**Comment by aadithya2112** [comment]
@coderabbitai full review and reevaluate all your previous comments that may have been dismissed, ignored or resolved. Evaluate whether the goals of the issue linked in the first PR comment are met. If there are test cases, ensure that the updated or created tests won't be flaky, especially in a sharded testing environment where mock cleanup is important and response times are highly variable. Post your conclusion as an easily readable comment that also includes your findings outside the diffs.

**Comment by aadithya2112** [comment]
@palisadoes This pr is ready for review

**Comment by aadithya2112** [comment]
@coderabbitai full review and reevaluate all your previous comments that may have been dismissed, ignored or resolved. Evaluate whether the goals of the issue linked in the first PR comment are met. If there are test cases, ensure that the updated or created tests won't be flaky, especially in a sharded testing environment where mock cleanup is important and response times are highly variable. Post your conclusion as an easily readable comment that also includes your findings outside the diffs.

**Comment by aadithya2112** [comment]
@palisadoes This pr is ready for review.

Note: I initially requested review before coderabbitai surfaced its suggested fixes (they appeared only after the report was generated), so I missed them at first. That’s now addressed.

File: schema.graphql [hunk 1]
```diff
@@ -5216,11 +5216,11 @@ type Query {
...
   user(input: QueryUserInput!): User
 
-  """Query field to fetch tags assigned to a user."""
+  """Fetch tags assigned to a specific user"""
   userTags(
-    """ID of the user whose tags are fetched"""
+    """User ID to fetch assigned tags for"""
     userId: ID!
-  ): [Tag!]!
+  ): [Tag!]
```
[the remark this answers is not in the document]
Comment (aadithya2112) [reply schema.graphql]: This is auto generated by the pre commit hook.

[the remark this answers is not in the document]
Comment (aadithya2112) [reply schema.graphql]: This appears to be a file that was not changed in my pr. This is probably due to my recent merge with develop. Please verify.

**Comment by aadithya2112** [comment]
@palisadoes Let me know what needs to be done regarding this.
https://github.com/PalisadoesFoundation/talawa-api/pull/5150#pullrequestreview-3766222125

I did fix and commit but I notice more issues from coderabbitai here https://github.com/PalisadoesFoundation/talawa-api/pull/5150#pullrequestreview-3766255202
These aren't related to my changes at all and looks to be existing issues from a pr recently.
Please check this comment from coderabbitai https://github.com/PalisadoesFoundation/talawa-api/pull/5150#discussion_r2776988858

**Comment by palisadoes** [comment]
@coderabbitai full review and reevaluate all your previous comments that may have been dismissed, ignored or resolved. Evaluate whether the goals of the issue linked in the first PR comment are met. If there are test cases, ensure that the updated or created tests won't be flaky, especially in a sharded testing environment where test cleanup is important and response times are highly variable. Post your conclusion as an easily readable comment that also includes your findings outside the diffs.
